// @build/pipelines/server.js
"use strict";

const core = require("../core");
const loader = core.getLoader();
const $ = loader.store;

const gulp = require("gulp");
const merge = require("merge-stream");

loader.depend({
	path: "path",

	gulpTypescript: "gulp-typescript",
	sourcemaps: "gulp-sourcemaps",
	gutil: "gulp-util",
	concat: "gulp-concat",
	insert: "gulp-insert",
	sort: "gulp-sort"
});

core.buildTasks.push("build:server");

/**
 * Prepares a string for use in a regular expression
 */
const escapeForRegEx = str => {
	return str.replace(/([\\*+?\[\].])/g, "\\$1");
};

/**
 * Sorts typings files so that the entrypoint is first.
 */
const sortTypings = pipeline => {
	const getRelevant = path => {
		const sourcedir = $.path.parse(pipeline.input).dir;
		const normalSource = $.path.normalize(sourcedir.replace(/\*/g, ""));
		const relevantMatcher = new RegExp(`^.+?${ escapeForRegEx(normalSource) }[/\\\\]?(.+)`);

		return path.match(relevantMatcher)[1];
	};

	const pipe = $.sort((a, b) => {
		let aR = getRelevant(a.path).replace(/\.d\.ts$/, ".ts");
		let bR = getRelevant(b.path).replace(/\.d\.ts$/, ".ts");

		if (aR === pipeline.moduleEntryPoint) {
			return 1;
		}

		if (bR === pipeline.moduleEntryPoint) {
			return -1;
		}

		return 0;
	});

	return pipe;
};

/**
 * A gulp transform that takes a list of .d.ts files and merges them.
 */
const mergeTypings = pipeline => {
	const sourcedir = $.path.parse(pipeline.input).dir;
	const normalSource = $.path.normalize(sourcedir.replace(/\*/g, ""));
	const moduleName = pipeline.moduleName;

	const pipe = $.insert.transform((contents, file) => {
		// Strip glob stars and normalize path

		// Find the relevant part of the path
		const relevantMatcher = new RegExp(`^.+?${ escapeForRegEx(normalSource) }[/\\\\]?(.+)`);
		const endName = file.path.match(relevantMatcher);

		if (!endName) {
			return contents;
		}

		// Strip the file extension off of the path
		const modulePath = endName[1].replace(/\.d\.ts$/, "");

		// Replace declarations like `export {default as X} from "x"`
		contents = contents.replace(/from "(.+?)"/g, (whole, path) => {
			const fullPath = $.path.join(moduleName, path);
			const normalized = $.path.normalize(fullPath)
				.replace(/\\/g, "/");

			return `from "${ normalized }"`;
		});

		// Indent blocks in to make them look nicer.
		contents = contents.replace(/\n/g, "\n\t").trim();

		let fileIdentifier = $.path.join(moduleName, modulePath)
			.replace(/\\/g, "/");

		if (modulePath === pipeline.moduleEntryPoint.replace(/\.[tj]s$/, "")) {
			fileIdentifier = moduleName;
		}

		return `declare module "${ fileIdentifier }" {\n\t${ contents }\n}`;
	});

	return pipe;
};

const buildPipeline = pipeline => {
	$.gutil.log($.gutil.colors.green(core.getName(pipeline) + ": building..."));

	const buildPath = core.get("out", pipeline);

	const dest = $.path.join(buildPath, pipeline.output);
	const sourcedir = $.path.parse(pipeline.input).dir;

	const entries = [pipeline.input];

	if (pipeline.extraEntries) {
		entries.push(...pipeline.extraEntries);
	}

	const config = {
		declaration: false
	};

	if (pipeline.typingsDest) {
		config.declaration = true;
	}

	if (pipeline.config && pipeline.config.typescript) {
		Object.assign(config, pipeline.config.typescript);
	}

	const handler = core.getErrorHandler(core.getName(pipeline));

	let stream = gulp.src(entries)
		.pipe($.sourcemaps.init())
		.on("error", handler)
		.pipe($.gulpTypescript(config));

	const streams = [];

	let js = stream.js
		.pipe($.sourcemaps.write("./"))
		.pipe(gulp.dest(dest));

	streams.push(js);

	if (pipeline.typingsDest) {
		const outDir = $.path.join(buildPath, $.path.dirname(pipeline.typingsDest));
		const outFileName = $.path.basename(pipeline.typingsDest);

		const typingsDest = $.path.join(buildPath, pipeline.typingsDest);

		let dts = stream.dts
			.pipe(sortTypings(pipeline))
			.pipe(mergeTypings(pipeline))
			.pipe($.concat(outFileName))
			.pipe(gulp.dest(outDir));

		streams.push(dts);
	}

	stream = merge(streams)
		.pipe(core.getNotify(core.getName(pipeline) + ": done!"));

	return stream;
};

gulp.task("build:server", () => {
	const pipelines = core.getPipelines("server");
	const merged = merge();

	if (pipelines.length > 0) {
		loader.load();
	}

	for (const pipeline of pipelines) {
		merged.add(buildPipeline(pipeline));

		if (core.get("watch", pipeline)) {
			gulp.watch(pipeline.input, e => {
				buildPipeline(pipeline);
			});
		}
	}

	return merged;
});
