// @build/args.js
"use strict";

const core = require("./core");

const argTypes = {
	watch: "boolean",
	production: "boolean",
	sourcemaps: "boolean",
	minify: "boolean",
	notify: "boolean",
	browsersync: "boolean",

	white: "set",
	gray: "set",
	black: "set",

	preset: "string",
	out: "string"
};

const aliases = {
	"only": "white",
	"except": "black"
};

const toBoolean = (v) => {
	v = v.toLowerCase();

	if (v === "true" || v === "yes") {
		return true;
	}

	if (v === "false" || v === "no") {
		return false;
	}

	return true;
};

module.exports = (cmdArgs) => {
	// pull the process arguments we care about out
	let args = [];

	for (let i = 2; i < cmdArgs.length; i++) {
		const arg = cmdArgs[i];

		if (arg.startsWith("--")) {
			args = cmdArgs.slice(i);
			break;
		}
	}

	// black list: don't build these modules
	// white list: only build these modules
	// gray list: build these modules despite defaults
	let params = core.params = {
		white: null,
		gray: null,
		black: null
	};

	for (let arg of args) {
		if (!arg.startsWith("--")) {
			console.error(`Unknown command line option '${ arg }'`);
			process.exit(1);
		}

		// Normalize and alias arguments
		arg = arg.slice(2).toLowerCase();
		arg = aliases[arg] || arg;

		let setter = arg.match(/^([^=]+)=(.+)/);

		// Flag, not a setter!
		if (!setter) {
			if (argTypes[arg] === "boolean") {
				params[arg] = true;
				continue;
			}

			console.error(`Unknown command line option '${arg}'`);
			process.exit(1);
		}

		let key = setter[1];
		let value = setter[2];

		key = aliases[key] || key;

		if (argTypes[key] === "set") {
			if (params[key] == null) {
				params[key] = {};
			}

			let list = value.split(",");
			for (let e of list) {
				params[key][e.toLowerCase()] = true;
			}

			continue;
		}


		if (argTypes[key] === "boolean") {
			params[key] = toBoolean(value);
			continue;
		}

		if (argTypes[key] === "string") {
			params[key] = value;
			continue;
		}
	}
};
