// @build/core.js
"use strict";

// load configuration
let conf;
try {
	conf = require("../../build.user.conf");
} catch(e) {
	conf = require("../../build.conf");
}

// pull the process arguments we care about out
let args = [];

for (let i = 2; i < process.argv.length; i++) {
	let arg = process.argv[i];

	if (arg.startsWith("--")) {
		args = process.argv.slice(i);
		break;
	}
}

// black list: don't build these modules
// white list: only build these modules
// gray list: build these modules despite defaults
let params = {
	white: null,
	gray: null,
	black: null
};

{
	let isList = false;
	let listMode = "";

	let modes = {
		white: true,
		gray: true,
		black: true
	};

	let aliases = {
		"only": "white",
		"except": "black"
	};

	for (let arg of args) {
		if (arg.startsWith("--")) {
			arg = arg.slice(2);

			arg = aliases[arg] || arg;

			if (modes[arg]) {
				isList = true;
				listMode = arg;

				if (params[arg] == null) {
					params[arg] = {};
				}
			}
		} else {
			if (isList) {
				let list = arg.split(",");
				for (let e of list) {
					params[listMode][e.toLowerCase()] = true;
				}
			}
		}
	}
}

let getValue = (name, base, core) => {
	let conf = core.conf;

	if (core.flags[name] != null) {
		return core.flags[name];
	}

	if (conf[name] != null) {
		return conf[name];
	}

	return base;
};

let core = {
	version: "2.0.0",

	modules: ["browser", "server", "styles", "static"],
	buildTasks: [],
	browserSync: null,

	conf: conf,
	params: params,
	flags: {
		once: null,
		production: null
	},

	get once() {
		return getValue("once", false, core);
	},

	get production() {
		return getValue("production", false, core);
	},

	get sourcemaps() {
		return getValue("sourcemaps", !core.production, core);
	},

	get minify() {
		return getValue("minify", !core.production, core);
	},

	load: (path) => {
		try {
			return require(path);
		} catch(e) {
			console.log(`Couldn't load module "${path}", try running "npm install"`);
			process.exit();
		}
	},

	getBuildPath: () => core.production ? core.conf.releasePath : core.conf.debugPath,

	getName: (transform) => {
		if (transform.name) {
			return transform.name;
		}

		return transform.type.charAt(0).toUpperCase() + transform.type.slice(1);
	},

	getTransforms: (type) => {
		let params = core.params;

		return core.conf.transforms.filter(transform => {
			let name = (transform.name || transform.type).toLowerCase();

			if (type != null && transform.type !== type) {
				return false;
			}

			if (params.black && params.black[name]) {
				return false;
			}

			if (params.white && !params.white[name]) {
				return false;
			}

			if (transform.disabled) {
				if (params.gray) {
					if (!params.gray[name]) {
						return false;
					}
				} else {
					return false;
				}
			}

			return true;
		});
	},

	onError: function(err) {
		require("gulp-notify").onError("Error: <%= error.message %>")(err);

		this.emit("end");
	}
};

module.exports = core;
